lsp/nls/src/trace.rs

```diff
@@ -0,0 +1,262 @@
+}
+
+impl Writer for Trace {
+    fn write_item(&mut self, item: TraceItem<Replied>) -> Result<()> {
+        if let Some(writer) = self.writer.as_mut() {
+            writer.write_item(item)
+        } else {
+            Ok(())
+        }
+    }
+
+    fn flush(&mut self) -> Result<()> {
+        if let Some(writer) = self.writer.as_mut() {
+            writer.flush()
+        } else {
+            Ok(())
+        }
+    }
+}
+
+impl Trace {
+    pub fn set_writer<W>(writer: W) -> Result<()>
+    where
+        W: Writer + Send + 'static,
+    {
+        Self::with_trace(|mut trace| {
+            let old = trace.writer.replace(Box::new(writer));
+            if let Some(mut writer) = old {
+                writer.flush()?;
+            }
+            Ok(())
+        })
+    }
+
+    #[allow(dead_code)]
+    pub fn remove_writer() -> Result<()> {
+        Self::with_trace(|mut trace| {
+            if let Some(ref mut writer) = trace.writer {
+                writer.flush()?;
+            }
+            trace.writer = None;
+            Ok(())
+        })
+    }
+
+    fn with_trace<F>(f: F) -> anyhow::Result<()>
+    where
+        F: FnOnce(MutexGuard<Trace>) -> Result<()>,
+    {
+        return TRACE
+            .lock()
+            .or_else(|_| anyhow::bail!("Could not lock tracer mutex"))
+            .and_then(f);
+    }
+
+    pub fn receive(id: RequestId, method: impl ToString) {
+        // let lock not affect receive time
+        let time = Received(Instant::now());
+
+        Self::with_trace(|mut trace| {
+            let params = TraceItemParams {
+                method: method.to_string(),
+                ..Default::default()
+            };
+            trace.received.insert(id, TraceItem { time, params });
+            Ok(())
+        })
+        .report();
+    }
+
+    pub fn reply_with(id: RequestId, with_error: bool) {
+        Self::with_trace(|mut t| {
+            t.received
+                .remove(&id)
+                .map(|received| received.into_replied(with_error))
+                .map(|item| t.write_item(item))
+                .unwrap_or(Ok(()))
+        })
+        .report()
+    }
+
+    pub fn reply(id: RequestId) {
+        Self::reply_with(id, false)
+    }
+
+    pub fn error_reply(id: RequestId) {
+        Self::reply_with(id, true)
+    }
+
+    #[allow(dead_code)]
+    pub fn drop_received() -> anyhow::Result<()> {
+        Self::with_trace(|mut t| {
+            t.received.clear();
+            Ok(())
+        })
+    }
+}
+
+pub(crate) trait Enrich<E: Enrichment> {
+    fn enrich(id: &RequestId, param: E);
+}
+
+trait ResultExt<E> {
+    fn report(&self);
+}
+
+impl<T, E: Display> ResultExt<E> for Result<T, E> {
+    fn report(&self) {
+        if let Err(e) = self {
+            error!("{}", e);
+        };
+    }
+}
+
+pub mod param {
+
+    use super::{Enrich, ResultExt, Trace};
+    use lsp_server::RequestId;
+    use nickel::typecheck::linearization::Completed;
+
+    pub trait Enrichment {}
```

Comment:
What is the role of `Enrichment` with respect to `Enrich<T>` ? Could we not just use `Enrich<T>` constraints instead of the additional `Enrichment`?

Reply:
Yes it's a bit of a useless marker trait here
I started off with something more generic but still typesafe in mind but settled with the current impl which at least  works for our use-case.

Reply:
I think you could literally replace `Enrichment` with `Enrich` and the `E` parameter of `Enrich` by `Self`? It's not very important but if it is the case it's probably better to remove superfluous code.